msteams-platform/toolkit/toolkit-v4/debug-your-Teams-app-test-tool-vs.md

```diff
@@ -0,0 +1,227 @@
+| Trigger Conversation Update Activity | Add user <br><br> Add bot <br><br> Add channel | `onMembersAdded`<br><br> `onTeamsMembersAddedEvent` <br><br> `onTeamsChannelRenamedEvent` |
+| | Remove user <br><br><br> Remove bot <br><br><br> Remove channel <br><br> Remove team | `onMembersRemoved` <br> `onTeamsMembersRemovedEvent` <br><br> `onMembersRemoved` <br> `onTeamsMembersRemovedEvent` <br><br> `onTeamsChannelDeletedEvent` <br><br> `onTeamsTeamDeletedEvent` |
+| | Rename channel <br><br> Rename team | `onTeamsChannelRenamedEvent` <br><br> `onTeamsTeamRenamedEvent` |
+
+> [!NOTE]
+> All types of activities aren't available in all scopes. For example, you can't add or remove a channel in a personal chat or a group chat.
+
+Predefined activity triggers are available in the **Mock an Activity** menu in Test Tool.
+
+To mock an **Add user** activity, follow these steps:
+
+1. In Visual Studio Code, go to **Solution Explorer**.
+1. Select **Program.cs** file.
+1. In the **Program.cs** file under `builder.Services.AddTransient<IBot>(sp =>`, add the following code:
+
+    
+    ```json
+    app.OnConversationUpdate("membersAdded", async (context, state, cancellationToken) =>
+    {
+    await context.SendActivityAsync($"new member added", cancellationToken: cancellationToken);
+    });
+    ```
+
+   :::image type="content" source="../../assets/images/teams-toolkit-v2/debug-VS/add-a-user-code-vs.png" alt-text="Screenshot shows the code added to program.cs file for predefined mock activity add user.":::
+
+1. In the Test Tool, go to **Mock an Activity** and select **Add user**.
+
+   :::image type="content" source="../../assets/images/teams-toolkit-v2/debug-VS/add-user-vs.png" alt-text="Screenshot shows the add user option under mock an activity.":::
+
+   A pop-up window appears to preview the activity handler.
+
+1. Select **Send activity**.
+
+   :::image type="content" source="../../assets/images/teams-toolkit-v2/debug-VS/add-a-user-request-vs.png" alt-text="Screenshot shows the option to send activity for predefined mock activity add user.":::
+
+   Bot sends a response.
+
+   :::image type="content" source="../../assets/images/teams-toolkit-v2/debug-VS/add-a-user-response-vs.png" alt-text="Screenshot shows the response of predefined mock activity add user.":::
+
+### Custom activity triggers
+
+You can use **Custom activity** to customize activity triggers such as, `reactionsAdded`, to fit the requirements of your bot app. Test Tool automatically populates the required properties of the activity. You can also modify the activity type and add more properties.
+
+1. In Visual Studio Code, go to **Solution Explorer**.
+1. Select **Program.cs** file.
+1. In the **Program.cs** file under `builder.Services.AddTransient<IBot>(sp =>`, add the following code:
+
+    
+    ```json
```

Comment:
same here, update the code snippet language to `csharp` and update the code indent.

Reply:
Updated.